### PR TITLE
README update

### PR DESCRIPTION
**Pull request by oliver-walker-hms**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -172,6 +178,131 @@ The OSIsoft connector now has the ability to support OMF. To enable OMF follow t
    1. IMPORTANT: The PI Data Archive Server can disconnect for various reasons. Make sure this is connected if you have a connection issue.
    1. Click confirm and finish running the utility.
 
+## OSIsoft Cloud Service
+
+### OCS Account Setup
+The OSIsoft connector supports an OCS connection. To make use of this feature follow the below steps.
```
Comment (alexjhawk) [inline README.md:184]:
```suggestion
The OSIsoft connector supports an OCS connection. To make use of this feature, follow the below steps.
```

File: README.md [hunk 2]
```diff
@@ -172,6 +178,131 @@ The OSIsoft connector now has the ability to support OMF. To enable OMF follow t
...
+   1. Select the client you just created and hit next.
+   1. Select the namespace you created and hit next.
+   1. Click the "Save" button.
+   1. Expand the dropdown menu that now shows up under the list of clients.
+   1. Click on your new client, more details will be available on the right hand panel.
+1. Obtain the Client ID.
+   1. Navigate to the namespace tab
+   1. Select your namespace
+   1. Click the "Display Details" button.
+   1. Copy the Account id for later use. This is your tenant ID.
```
Comment (alexjhawk) [inline README.md:211]: should be properly numbered

File: README.md [hunk 3]
```diff
@@ -172,6 +178,131 @@ The OSIsoft connector now has the ability to support OMF. To enable OMF follow t
...
+### Flexy OCS Configuration File Changes
+1. Add OCS information to the ConnectorConfig.json file
+   1. Replace the xxx below with your namespace
+   1. Replace the yyy below with your tenant ID 
+   1. Copy these into the ServerConfig portion of the json file
+      ```
+      "namespace":"xxx",
+      "tenantId":"yyy"
+      ```
+   1. Make sure that there is a comma at the end of the line if it is not the last line in the section. An example of a properly written ConnectorConfig file is below
```
Comment (alexjhawk) [inline README.md:225]: same comment about proper numbering

File: README.md [hunk 4]
```diff
@@ -172,6 +178,131 @@ The OSIsoft connector now has the ability to support OMF. To enable OMF follow t
...
+
+1. Copy the below basic script to the Init Section of the Flexy Basic IDE
+   1. On the Flexy's web page, click to expand the "Setup" section.
+   1. Click on the "BASIC IDE" tab.
+   1. Copy and paste the below basic script into the Init section
+   1. Change the xxxxxx for client ID to be your URL encoded client id that you saved previously.
+   1. Change the yyyyyy for client secret your URL encoded client secret you saved previously.
+   1. Click on "File"
+   1. Click on "Save"
+   1. Click on the script execution slider that says "Stopped" to switch it to say "Running"
```
Comment (alexjhawk) [inline README.md:280]: proper numbering

File: README.md [hunk 5]
```diff
@@ -22,9 +22,15 @@ This application is supported by HMS' North American offices.
    3. [Installation](#Installation)
    4. [Running](#Running)
 2. [Certificates](#Certificates)
+   1. [Generate the certificate](#generate-the-certificate)
+   2. [Install the certificate on server](#install-the-certificate-on-server)
+   3. [Transfer certificate to Flexy](#transfer-certificate-to-flexy)
```
Comment (TomKimsey) [inline README.md:27]: Mixing capitalization

File: README.md [hunk 6]
```diff
@@ -22,9 +22,15 @@ This application is supported by HMS' North American offices.
...
 3. [Configuration File](#Configuration-File)
-4. [Tag Configuration](#Tag-Configuration)
-5. [Customizing the Application](#Customizing-the-application)
+4. [Customizing the Application](#Customizing-the-application)
+5. [Tag Configuration](#Tag-Configuration)
+6. [OMF support](#OMF-Support)
```
Comment (TomKimsey) [inline README.md:31]: capitalize Support

File: README.md [hunk 7]
```diff
@@ -22,9 +22,15 @@ This application is supported by HMS' North American offices.
...
+7. [OSIsoft Cloud Service](#OSIsoft-Cloud-Service)
+8. [Troubleshooting common issues](#Troubleshooting-common-issues)
```
Comment (TomKimsey) [inline README.md:33]: Capitalization

File: README.md (hunk 2 above)
Comment (oliver-walker-hms) [reply README.md:211]: it changes 1 to whatever the correct number is

File: README.md [hunk 8]
```diff
@@ -172,6 +178,131 @@ The OSIsoft connector now has the ability to support OMF. To enable OMF follow t
...
+      ```
+   1. Make sure that there is a comma at the end of the line if it is not the last line in the section. An example of a properly written ConnectorConfig file is below
+   ```
+   {
+   "ServerConfig":{
+      "IP":"192.168.1.8",
+      "WebID":"F1DS4knnwtuIDUC1RLy6XJGV4QREVTS1RPUC0wQ1MyUDFT",
+      "Credentials":"dsZtOnBfdsjklfh==",
+      "namespace":"testNameSpace1",
+      "tenantId":"4da89f7d-b3ce-48a6-b906-b96c367000a9"
```
Comment (TomKimsey) [inline README.md:233]: Why are these not capitalized?

File: README.md (hunk 4 above)
Comment (TomKimsey) [inline README.md:280]: Some of these end in period, some do not.

File: README.md [hunk 9]
```diff
@@ -172,6 +178,131 @@ The OSIsoft connector now has the ability to support OMF. To enable OMF follow t
...
+  ClientSecret$ = "yyyyyy"
+  DataToSend$ = "client_id="+ ClientId$ +"&client_secret="+ ClientSecret$+"&grant_type=client_credentials"
+  File$ = "/usr/token.json"
+
+  REQUESTHTTPX URL$, "POST",Header$ , DataToSend$, "", File$
+
+  // change tag value here to mark as set
+  tokenReq@ = (tokenReq@ + 1)
+ENDFN
+```
```
Comment (TomKimsey) [inline README.md:302]: Shouldnt this be a file added to the release zip?

File: README.md (hunk 1 above)
Comment (oliver-walker-hms) [reply README.md]: fixed

File: README.md [hunk 10]
```diff
@@ -172,6 +178,131 @@ The OSIsoft connector now has the ability to support OMF. To enable OMF follow t
...
+
+### OCS Account Setup
+The OSIsoft connector supports an OCS connection. To make use of this feature, follow the below steps.
+1. First create an OCS account.
+1. Under the namespace tab, create an OCS namespace of your choosing.
+   1. Save the namespace for later use. It will be needed in another step.
+1. Under the client tab, create a client.
+   1. Select "Client-Credentials" as the client type.
+   1. Click the "Add Client button.
+   1. Give the client a name. NOTE: the client will be the flexy device.
```
Comment (alexjhawk) [inline README.md:191]:
```suggestion
   1. Give the client a name. NOTE: the client will be the Flexy device.
```

File: README.md (hunk 9 above)
Comment (oliver-walker-hms) [reply README.md:302]: It's pretty short were copy and paste is just as easy as the import. I suppose either way would be fine. Should i move this and suggest they import it into the basic ide?

Comment (TomKimsey) [reply README.md:302]: I think its a lot safer and to have them import it from a file. That way you eliminate the chance for copy/paste errors.

Comment (TomKimsey) [reply README.md:302]: This chunk of code also needs to be formatted, a lot of inconstant spacing.

File: README.md [hunk 11]
```diff
@@ -172,6 +178,131 @@ The OSIsoft connector now has the ability to support OMF. To enable OMF follow t
...
+The Client ID and the Client Secret will be needed for the basic script you will have running on the Flexy.
+
+### Flexy OCS Configuration File Changes
+1. Add OCS information to the ConnectorConfig.json file
+   1. Replace the xxx below with your namespace
+   1. Replace the yyy below with your tenant ID 
+   1. Copy these into the ServerConfig portion of the json file
+      ```
+      "namespace":"xxx",
+      "tenantId":"yyy"
```
Comment (alexjhawk) [inline README.md:223]: should be capitalized to match below

**Comment by oliver-walker-hms** [comment]
I tossed in another random commit that was going to be it's own pr to fix something I messed up in the last PR

File: scripts/OcsBasicScript.txt [hunk 12]
```diff
@@ -0,0 +1,27 @@
+Rem --- eWON start section: Cyclic Section
+eWON_cyclic_section:
+Rem --- eWON user (start)
+
+Rem --- eWON user (end)
+End
+Rem --- eWON end section: Cyclic Section
+Rem --- eWON start section: Init Section
+eWON_init_section:
+Rem --- eWON user (start)
```
Comment (TomKimsey) [inline scripts/OcsBasicScript.txt:10]: Ewon capitalization

File: scripts/OcsBasicScript.txt [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+
+Rem --- eWON user (end)
+End
+Rem --- eWON end section: Cyclic Section
+Rem --- eWON start section: Init Section
+eWON_init_section:
+Rem --- eWON user (start)
+// Timer 60 seconds * 30 minutes
+TSET 1, (60 * 30)
```
Comment (TomKimsey) [inline scripts/OcsBasicScript.txt:12]: Why not use 1800 here?

Comment (TomKimsey) [reply scripts/OcsBasicScript.txt:12]: And have it be a variable to remove the magic number

File: scripts/OcsBasicScript.txt [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+ONTIMER 1, "@updateToken()"
+END
+
+Function updateToken()
+  URL$ = "https://dat-b.osisoft.com/identity/connect/token"
+  Header$ = ""
+  ClientId$ = "6126b79f-458a-49e9-b306-767d1ee1a1cb"
+  ClientSecret$ = "6JafF8HtDp3RULj9SGP4l7eoAvP2%2FcwHQaniB0G5nqw%3D"
```
Comment (TomKimsey) [inline scripts/OcsBasicScript.txt:20]: Are these fields not unique to your setup and should be shared in the repo?

File: src/com/hms/flexyosisoftconnector/OSIsoftServer.java [hunk 15]
```diff
@@ -262,6 +262,11 @@ public int initTags() throws JSONException {
     int res;
     final String responseFilename = "/usr/response.json";
 
+    // set batch buffer with list of tag containers to initialize for if they have not been
+    // already
```
Comment (TomKimsey) [inline src/com/hms/flexyosisoftconnector/OSIsoftServer.java:266]: Multiline comment should not use //

File: README.md [hunk 16]
```diff
@@ -172,6 +178,134 @@ The OSIsoft connector now has the ability to support OMF. To enable OMF follow t
...
+  ClientId$ = "6126b79f-458a-49e9-b306-767d1ee1a1cb"
+  ClientSecret$ = "6JafF8HtDp3RULj9SGP4l7eoAvP2%2FcwHQaniB0G5nqw%3D"
+  DataToSend$ = "client_id=" + ClientId$ + "&client_secret=" + ClientSecret$ + "&grant_type=client_credentials"
+  File$ = "/usr/token.json"
+  REQUESTHTTPX URL$, "POST", Header$, DataToSend$, "", File$
+
+  // change tag value here to mark as set
+  tokenReq@ = (tokenReq@ + 1)
+ENDFN
+```
```
Comment (TomKimsey) [inline README.md:305]: why is this still here since the readme has the instructions to import the file?

File: scripts/OcsBasicScript.txt [hunk 17]
```diff
@@ -0,0 +1,27 @@
+Rem --- eWON start section: Cyclic Section
```
Comment (TomKimsey) [inline scripts/OcsBasicScript.txt:1]: Why is this not .bas extension?

Comment (oliver-walker-hms) [reply scripts/OcsBasicScript.txt:1]: ah, that's what it was. It seems to not even look at the extension but that's good to know, will change

File: README.md (hunk 16 above)
Comment (oliver-walker-hms) [reply README.md:305]: I figured this would be a good check since people get easily confused.

Comment (oliver-walker-hms) [reply README.md:305]: I guess you're right and it's just clutter now. I can remove this

File: README.md [hunk 18]
```diff
@@ -172,6 +178,113 @@ The OSIsoft connector now has the ability to support OMF. To enable OMF follow t
...
+   }
+   ```
+   1. Upload the updated ConnectorConfig.json file to the /usr directory of your Flexy.
+
+### Flexy OCS BASIC Script
+1. Create a tag in your Flexy for the Basic script to store information.
+   1. On the Flexy's web page, click to expand the "Tags" section.
+   1. Click the "Values" tab.
+   1. Click the "MODE" slider to say "setup".
+   1. Click the "add" button.
```
Comment (TomKimsey) [inline README.md:253]: is add lowercase?

File: README.md [hunk 19]
```diff
@@ -172,6 +178,113 @@ The OSIsoft connector now has the ability to support OMF. To enable OMF follow t
...
+### Flexy OCS Configuration File Changes
+1. Add OCS information to the ConnectorConfig.json file.
+   1. Replace the xxx below with your namespace.
+   1. Replace the yyy below with your tenant ID.
+   1. Copy these into the ServerConfig portion of the json file.
+      ```
+      "Namespace":"xxx",
+      "TenantId":"yyy"
+      ```
+   1. Make sure that there is a comma at the end of the line if it is not the last line in the section. An example of a properly written ConnectorConfig file is below.
```
Comment (TomKimsey) [inline README.md:225]: This line is a bit confusing as it attempts to talk about JSON formatting, yet says "is a comma at the end of the line if it is not the last line in the section" which does not really make sense.

Comment (oliver-walker-hms) [reply README.md]: I change it to: Make sure that the JSON formatting is valid. There are multiple online tools to check this. Below is an example of a properly formatted configuration file.

File: README.md (hunk 18 above)
Comment (oliver-walker-hms) [reply README.md]: fixed, the button has it as Uppercase

File: README.md [hunk 20]
```diff
@@ -172,6 +178,113 @@ The OSIsoft connector now has the ability to support OMF. To enable OMF follow t
...
+### Flexy OCS Configuration File Changes
+1. Add OCS information to the ConnectorConfig.json file.
+   1. Replace the xxx below with your namespace.
+   1. Replace the yyy below with your tenant ID.
+   1. Copy these into the ServerConfig portion of the json file.
+      ```
+      "Namespace":"xxx",
+      "TenantId":"yyy"
+      ```
+   1. Make sure that the JSON formatting is valid. There are multiple online tools to check this. Below is an example of a properly formatted configuration file.
```
Comment (TomKimsey) [inline README.md:225]:
```suggestion
   1. Make sure that the JSON formatting is valid. Below is an example of a properly formatted configuration file.
```

Comment (TomKimsey) [reply README.md:225]: They should get an error if the formatting is not right. Its a little weird to suggest that tools exist and not give an example. If someone messes up the formatting they should be able to figure it out.

**Pull request merged**